provisioning/baremetal_config.go

```diff
@@ -74,58 +121,121 @@ func getProvisioningNetworkMode(prov *metal3iov1alpha1.Provisioning) metal3iov1a
+		// If it's not a valid URI lets just return.
+		return errs
+	}
+	if parsedURL.Scheme != "http" && parsedURL.Scheme != "https" {
+		errs = append(errs, fmt.Errorf("the provisioningOSDownloadURL provided: %s must begin with http/https", uri))
+		// Again it's not worth it if it's not http(s)
+		return errs
+	}
+	var sha256Checksum string
+	if sha256Checksums, ok := parsedURL.Query()["sha256"]; ok {
+		sha256Checksum = sha256Checksums[0]
+	}
+	if sha256Checksum == "" {
+		errs = append(errs, fmt.Errorf("the sha256 parameter in the provisioningOSDownloadURL %q is missing", uri))
+	}
+	if len(sha256Checksum) != 64 {
+		errs = append(errs, fmt.Errorf("the sha256 parameter in the provisioningOSDownloadURL %q is invalid", uri))
+	}
+	if !strings.HasSuffix(parsedURL.Path, ".qcow2.gz") && !strings.HasSuffix(parsedURL.Path, ".qcow2.xz") {
+		errs = append(errs, fmt.Errorf("the provisioningOSDownloadURL provided: %q is an OS image and must end in .qcow2.gz or .qcow2.xz", uri))
+	}
+
+	return errs
 }
 
-func validateUnmanagedConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningInterface", Value: prov.Spec.ProvisioningInterface},
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
-		}
+func validateProvisioningInterface(iface string) []error {
+	// Unfortunately there seems to be nothing you can really do to verify
+	// the ProvisioningInterface as it's not necessarily on this machine
+	// and could be almost anything.
+	var errs []error
+
+	if iface == "" {
+		errs = append(errs, fmt.Errorf("provisioningInterface is required but is not set"))
 	}
-	return nil
+	return errs
 }
 
-func validateDisabledConfig(prov *metal3iov1alpha1.Provisioning) error {
-	for _, toTest := range []struct {
-		Name  string
-		Value string
-	}{
-
-		{Name: "ProvisioningIP", Value: prov.Spec.ProvisioningIP},
-		{Name: "ProvisioningNetworkCIDR", Value: prov.Spec.ProvisioningNetworkCIDR},
-		{Name: "ProvisioningOSDownloadURL", Value: prov.Spec.ProvisioningOSDownloadURL},
-	} {
-		if toTest.Value == "" {
-			return fmt.Errorf("%s is required but is empty", toTest.Name)
+func validateProvisioningNetworkSettings(ip string, cidr string, dhcpRange string) []error {
+	// provisioningIP and networkCIDR are always set.  DHCP range is optional
+	// depending on mode.
+	var errs []error
+
+	// Verify provisioning ip and get it into net format for future tests.
+	provisioningIP := net.ParseIP(ip)
+	if provisioningIP == nil {
+		errs = append(errs, fmt.Errorf("The provisioningIP %q is not a valid IP", ip))
+		// I think we return here.  There are a few more tests below that require the IP be set.
```

Comment:
Is this a comment from when the PR was WIP?

Reply:
It's still valid but I'll remove it.